influxql/iterator.gen.go

```diff
@@ -956,6 +956,90 @@ func (itr *floatReaderIterator) Next() *FloatPoint {
 	return p
 }
 
+// floatMeanIterator returns the mean of values for every interval.
+type floatMeanIterator struct {
+	input  *bufFloatIterator
+	opt    IteratorOptions
+	points []*FloatPoint
+}
+
+// Close closes the iterator and all child iterators.
+func (itr *floatMeanIterator) Close() error { return itr.input.Close() }
+
+// Next returns the next value.
+func (itr *floatMeanIterator) Next() *FloatPoint {
+	// Calculate the next window if we have no more points.
+	if len(itr.points) == 0 {
+		itr.points = itr.next()
+		if len(itr.points) == 0 {
+			return nil
+		}
+	}
+
+	// Pop next point off the stack.
+	p := itr.points[len(itr.points)-1]
+	itr.points = itr.points[:len(itr.points)-1]
+	return p
+}
+
+func (itr *floatMeanIterator) next() []*FloatPoint {
+	startTime, endTime := itr.opt.Window(itr.input.peekTime())
+
+	// Create points by tags.
+	m := make(map[string]struct {
+		Name  string
+		Tags  Tags
+		Sum   float64
+		Count uint32
+	})
+	for {
+		// Read next point.
+		curr := itr.input.NextInWindow(startTime, endTime)
+		if curr == nil {
+			break
+		} else if curr.Nil {
+			continue
+		}
+		tags := curr.Tags.Subset(itr.opt.Dimensions)
+		id := curr.Name + "\x00" + tags.ID()
+
+		prev, ok := m[id]
+		if !ok {
+			prev.Name = curr.Name
+			prev.Tags = tags
+		}
+
+		if curr.Aggregated != 0 {
+			prev.Sum += curr.Value * float64(curr.Aggregated)
+			prev.Count += curr.Aggregated
+		} else {
+			prev.Sum += curr.Value
+			prev.Count++
+		}
```

Comment:
Specifically this code could be placed in a function so that Kapacitor can reuse.

This bring up Map/Reduce in general. Before the map phase was allowed to produce a custom struct that could contain intermediate results for the reduce phase to consume. With this new framework this is not possible. Are we going to produce custom iterators for those use cases? Seems like a more extensible design would be to allow for custom structure between the phases.